api.planx.uk/modules/flows/validate/service.ts

```diff
@@ -206,4 +211,63 @@ const inviteToPayEnabled = (flowGraph: FlowGraph): boolean => {
   );
 };
 
+const validateFileTypes = (flowGraph: FlowGraph): ValidationResponse => {
+  // Get all passport variables set by FileUpload and/or FileUploadAndLabel
+  const allFileFns = [
+    ...getFileUploadNodeFns(flowGraph),
+    ...getFileUploadAndLabelNodeFns(flowGraph),
+  ];
+  if (allFileFns.length < 1) {
+    return {
+      title: "File types",
+      status: "Not applicable",
+      message: "Your flow is not using FileUpload or UploadAndLabel",
+    };
+  }
+
+  // Get all file types supported by current release of ODP Schema & compare
+  const validFileTypes = getValidSchemaValues("FileType");
+  const invalidFileFns: string[] = [];
+  allFileFns.forEach((fn) => {
+    if (!validFileTypes?.includes(fn)) {
+      invalidFileFns.push(fn);
+    }
+  });
+  if (invalidFileFns.length > 0) {
+    return {
+      title: "File types",
+      status: "Warn",
+      message: `Your FileUpload or UploadAndLabel are setting data fields that are not supported by the ODP Schema: ${invalidFileFns.join(", ")}`,
+    };
+  }
+
+  return {
+    title: "File types",
+    status: "Pass",
+    message:
+      "Files collected via FileUpload or UploadAndLabel are all supported by the ODP Schema",
+  };
+};
+
+const getFileUploadNodeFns = (flowGraph: FlowGraph): string[] => {
+  const fileUploadNodes = Object.entries(flowGraph).filter(
+    (entry): entry is [string, Node] =>
+      isComponentType(entry, ComponentType.FileUpload),
+  );
+  return fileUploadNodes.map(([_nodeId, node]) => node.data?.fn as string);
+};
+
+const getFileUploadAndLabelNodeFns = (flowGraph: FlowGraph): string[] => {
+  // Exclude Upload & Label nodes used in "info-only" mode with a hidden dropzone
+  const uploadAndLabelNodes = Object.entries(flowGraph).filter(
+    (entry): entry is [string, Node] =>
+      isComponentType(entry, ComponentType.FileUploadAndLabel) &&
+      entry[1].data?.hideDropZone !== true,
+  );
+  const uploadAndLabelFileTypes = uploadAndLabelNodes
+    .map(([_nodeId, node]) => node.data?.fileTypes)
+    .flat();
+  return uploadAndLabelFileTypes?.map((file: any) => file?.fn as string);
```

Comment:
`any` is less than ideal here, but file is an editor-only type definition right now and not a particularly simple one to repeat because of the `conditions` !

Reply:
Yeah this is a fair call, no issues, and thanks for the explanation.

It would be very very meaningful to put some work into a typed Flow data structure (been on the wish list for a while). I'd be keen to have a start at this at some point!